com.unity.netcode.gameobjects/Runtime/Core/NetworkBehaviour.cs

```diff
@@ -181,14 +187,56 @@ internal void __endSendClientRpc(ref FastBufferWriter bufferWriter, uint rpcMeth
                         shouldSendToHost = true;
                         break;
                     }
+
+                    // Check to make sure we are sending to only observers, if not log an error.
+                    if (!NetworkObject.Observers.Contains(targetClientId) && logLevel >= LogLevel.Error)
+                    {
+                        NetworkLog.LogError(GenerateObserverErrorMessage(clientRpcParams, targetClientId));
+                    }
                 }
 
                 rpcWriteSize = NetworkManager.SendMessage(ref clientRpcMessage, networkDelivery, clientRpcParams.Send.TargetClientIdsNativeArray.Value);
             }
             else
             {
                 shouldSendToHost = IsHost;
-                rpcWriteSize = NetworkManager.SendMessage(ref clientRpcMessage, networkDelivery, NetworkManager.ConnectedClientsIds);
+                var observerAllocateCount = NetworkObject.Observers.Count();
+                // Exclude accounting for the host client id as the message is handled locally
+                if (shouldSendToHost && NetworkObject.Observers.Contains(NetworkManager.ServerClientId))
+                {
+                    observerAllocateCount--;
+                }
+
+                // Only if at least 1 connected client or more to send to
+                if (observerAllocateCount > 0)
+                {
+                    // Allocate native array large enough to hold the remaining observers' clientIds.
+                    var observerClientIds = new NativeArray<ulong>(observerAllocateCount, Allocator.Temp);
+                    // Since we are modifying the array and arrays within a using statement cannot be modified
+                    // placing a try and finally around this section to assure we don't leak memory if anything
+                    // throws an exception before disposing the array
+                    try
+                    {
+                        var observerEnumerator = NetworkObject.Observers.GetEnumerator();
+                        var observerCount = 0;
+                        while (observerEnumerator.MoveNext())
+                        {
+                            // Skip over the host
+                            if (IsHost && observerEnumerator.Current == NetworkManager.LocalClientId)
+                            {
+                                shouldSendToHost = true;
+                                continue;
+                            }
+                            observerClientIds[observerCount++] = observerEnumerator.Current;
```

Comment:
Rather than go to the trouble of allocating and building up the client ids, why not simply here in the while loop do:

```
NetworkManager.SendMessage(ref clientRpcMessage, networkDelivery, observerEnumerator.Current)
```

Reply:
After our internal discussion, I changed my approach to avoid the temporary memory allocation.